### DISCUSSION TITLE
How to make a deformable cuboid

### DISCUSSION THREAD

**Discussion by ZSciamma** [InteractiveComputerGraphics/PositionBasedDynamics#127] [Q&A]
Hello,

I am trying to make a deformable cuboid, and I would greatly appreciate any advice.

I am using this library to make a simple physics-based game. I would like to make deformable walls, so that a character can bump into a wall and cause the wall to ripple or bounce back. First, I am trying to make a simple deformable regular cuboid, but I have been having some trouble with this. Here is the process I am using:

1. Create a cuboid in Blender
2. Export cuboid from Blender to OBJ and PLY.
3. Use [TetGen](https://wias-berlin.de/software/index.jsp?id=TetGen&lang=1), with cube PLY file as input, to create ELE and NODE files (tetrahedral mesh). Use -q or -a option to limit maximum tetrahedron edge length/volume and so create a finer mesh.
4. Create a new json scene with a rigid body floor, and a single TetModel (the cuboid). Populate the model information as below:
```
// newScene.json
{
	"id": 1,
	"nodeFile": "../models/blender_cuboid.node",
	"eleFile": "../models/blender_cuboid.ele",
	"visFile": "../models/blender_cuboid.obj",
	"resolutionSDF": [1,1,1],
	"translation": [0,6,0],
	"rotationAxis": [0, 0, 1],
	"rotationAngle": 0.1,
	"scale": [1,1,1],
	"staticParticles": [],
	"restitution" : 0.0,
	"friction" : 0.3,
	"collisionObjectType": 5,
	"collisionObjectFileName": "",
          "collisionObjectScale": [1,1,1],
	"testMesh": 1
}
```
5. Build project with `make`
6. Load scene with `./SceneLoaderDemo resources/scenes/newScene.json`

Unfortunately, I've been having a few problems. The cuboid seems rigid and not really deformable, and it doesn't collide well with other objects (other rigid bodies clip through it or get stuck to it). I also tried a tet sphere through the same process, and for the sphere the collisions work well, but it also doesn't seem to deform. I have a couple of questions:

1. How do I make the cuboid deformable? I thought that by creating it as a TetModel it would automatically be deformable, but it seems very rigid. I created some armadillos alongside it with almost identical json parameters, and they seem much more deformable. If I understand correctly, the deformable-ness comes from the Tet constraints, which are initialized in SceneLoaderDemo.cpp:initTetModelConstraints(). Is it incorrect on principle to try to make a deformable cuboid because that shape can't deform, or is there some specific way to specify that the cuboid should be deformable? I suspected that the cuboid just didn't have enough particles and that the tet mesh wasn't fine enough, so I tried making an absurd large model with an extensive tet mesh that looks like it should deform (**pictured below**) - but that also seems rigid.
2. Do you know how the armadillo models (/data/models/armadillo.obj) were created? I have been creating models in Blender, but it seems to me that the armadillo visualization mesh (extracted from the OBJ file) is finer and more detailed than the armadillo physics tetrahedral mesh (extracted from the ELE and NODE files). Did you create the armadillo OBJ in modelling software and then somehow use TetGen to reduce the level of detail of the mesh, in order to have fewer particles for physics purposes? When I create models in blender and then use TetGen to increase the number of tetrahedra in the tet mesh, I end up with a tet mesh which is much more detailed than the original OBJ file - and so the viz mesh looks too simple to show the level of detail and deformable-ness available in the physics. 

My apologies for the simple questions - I am new to graphics and I am not sure why I haven't been able to get this to work. Incidentally, I haven't found a tutorial or guide for this library, but if you know of somewhere I can find this information or some research I should do first, I would love to take a look.

Thank you so much for all your help!

![large mesh example](https://user-images.githubusercontent.com/17149360/164355882-19161c75-46af-483f-b38d-8e7b2451a188.png)

**Comment by janbender** [comment]
One problem is that the resolution of the SDF is far too low: "resolutionSDF": [1,1,1],
Try with at least: "resolutionSDF": [20,20,20], or higher.

Otherwise I don't see a big problem. Can you post the complete scene file?
